R/gh_token.R

```diff
@@ -152,33 +168,182 @@ gh_auth <- function(token) {
+#' Enterprise.
+#'
+#' @param url Character vector of HTTP/HTTPS URLs. They don't have to be in the
+#'   API-specific form, although they can be.
 #' @return Character vector of suffixes.
 #'
 #' @seealso [gh_token()]
 #' @export
 #' @examples
-#' # The main GH site
+#' # main github.com site
 #' slugify_url("https://api.github.com")
+#' slugify_url("https://github.com")
 #'
-#' # A custom one
+#' # an instance of GitHub Enterprise
 #' slugify_url("https://github.acme.com")
+#' slugify_url("https://github.acme.com/api/v3")
 
-slugify_url <- function(url) {
+slugify_url <- function(url) {           # https://jane@github.uni.edu/api/v3
+  url <- get_baseurl(url)                # https://jane@github.uni.edu
+  url <- normalize_host(url)
+  x2 <- sub("^.*://([^/]*@)?", "", url)  # github.uni.edu
+  x3 <- gsub("[.]+", "_", x2)            # github_uni_edu
+  x4 <- gsub("[^-a-zA-Z0-9_]", "", x3)
+  toupper(x4)                            # GITHUB_UNI_EDU
+}
+
+get_baseurl <- function(url) {               # https://github.uni.edu/api/v3/
   if (!any(grepl("^https?://", url))) {
     stop("Only works with HTTP(S) protocols")
   }
-  x2 <- sub("^.*://([^/]*@)?", "", url)
-  x3 <- sub("/+$", "", x2)
-  x4 <- gsub("[./]+", "_", x3)
-  x5 <- gsub("[^-a-zA-Z0-9_]", "", x4)
-  toupper(x5)
+  prot <- sub("^(https?://).*$", "\\1", url) # https://
+  rest <- sub("^https?://(.*)$", "\\1", url) #         github.uni.edu/api/v3/
+  host <- sub("/.*$", "", rest)              #         github.uni.edu
+  paste0(prot, host)                         # https://github.uni.edu
+}
+
+normalize_host <- function(x) {
+  sub("api.github.com", "github.com", x)
+}
+
+get_hosturl <- function(url) {
+  url <- get_baseurl(url)
+  normalize_host(url)
+}
+
+get_apiurl <- function(url) {
+  host_url <- get_hosturl(url)
+  prot_host <- strsplit(host_url, "://", fixed = TRUE)[[1]]
+  if (is_github_dot_com(host_url)) {
+    paste0(prot_host[[1]], "://api.github.com")
+  } else {
+    paste0(host_url, "/api/v3")
+  }
+}
+
+is_github_dot_com <- function(url) {
+  url <- get_baseurl(url)
+  url <- normalize_host(url)
+  grepl("^https?://github.com", url)
+}
+
+# inlined from credentials and generalized to non-github.com URL
+# is_interactive() is an unexported helper defined in utils.R
+# morally, is rlang::is_interactive()
+set_github_pat2 <- function(api_url = default_api_url(),
+                            force_new = FALSE,
+                            validate = is_interactive(),
+                            verbose = validate) {
+  if (!can_load("credentials")) {
+    return(FALSE)
+  }
+
+  host_url <- get_hosturl(api_url)
+  prot_host <- strsplit(host_url, "://", fixed = TRUE)[[1]]
+  if (is_github_dot_com(api_url)) {
+    # what's currently in credentials
+    # pat_user <- Sys.getenv("GITHUB_PAT_USER", 'PersonalAccessToken')
+    # TODO: I propose we change the user env var to GITHUB_USER
+    user_env_var <- "GITHUB_PAT_USER"
```

Comment:
Why do we actually need the user? If we use a PAT, then no user is needed, right? If we get the auth info from the credential store, that'll tell us the user?

Reply:
See this part of the vignette:

https://github.com/r-lib/gh/blob/9d188d3739af6adf77d70b8324260a2b5d71ed2a/vignettes/Managing-Personal-Access-Tokens.Rmd#L171-L189

@jeroen has convinced me this is correct. In my previous PR (https://github.com/r-lib/credentials/pull/7), I also wanted to set it to the actual username. But then it will clobber a username/password user may have cached through normal command line git via https.

Using 'PersonalAccessToken' is what [Git Credential Manager Core](https://github.blog/2020-07-02-git-credential-manager-core-building-a-universal-authentication-experience/) does.

The `user` also provides a non-obvious and not-documented way to use more than 1 PAT for a given GitHub host.

Reply:
Yes, what I meant is that we do not need the env var at all, and we don't need to worry about the username at all.

The vignettes discusses to use the username for multiple credentials, but I don't think we need to worry about this, other then possible allowing a username in the URL.